src/frontend/circom/mod.rs

```diff
@@ -0,0 +1,229 @@
+
+// Convert the generic type of constraints from Pairing's ScalarField to PrimeField.
+pub fn convert_constraints_bigint_to_scalar<E, F>(
+    constraints: Constraints<E::ScalarField>,
+) -> Constraints<F>
+where
+    E: Pairing,
+    F: PrimeField,
+{
+    let convert_vec = |vec: ConstraintVec<E::ScalarField>| -> ConstraintVec<F> {
+        vec.into_iter()
+            .map(|(index, element)| {
+                // Convert element into BigInt, then to BigUint, then to the destination PrimeField
+                let bigint_e: <<E as Pairing>::ScalarField as PrimeField>::BigInt =
+                    element.into_bigint();
+                let generic_biguint: num_bigint::BigUint = bigint_e.into();
+                let f_element: F = F::from(generic_biguint);
+                (index, f_element)
+            })
+            .collect()
+    };
+
+    (
+        convert_vec(constraints.0),
+        convert_vec(constraints.1),
+        convert_vec(constraints.2),
+    )
+}
+
+// Extract constraints on Pairing's ScalarField from an .r1cs file
+// and convert them into constraints on PrimeField
+pub fn extract_constraints_from_r1cs<E, F>(r1cs_filepath: &Path) -> ExtractedConstraintsResult<F>
+where
+    E: Pairing,
+    F: PrimeField,
+{
+    // Open the .r1cs file and create a reader
+    let file = File::open(r1cs_filepath)?;
+    let reader = BufReader::new(file);
+
+    // Read the R1CS file, extract the constraints, and then convert them into new constraints on PrimeField
+    let r1cs_file = r1cs_reader::R1CSFile::<E>::new(reader)?;
+    let pub_io_len = (r1cs_file.header.n_pub_in + r1cs_file.header.n_pub_out) as usize;
+    let r1cs = r1cs_reader::R1CS::<E>::from(r1cs_file);
+    let converted_constraints: Vec<Constraints<F>> = r1cs
+        .constraints
+        .into_iter()
+        .map(|c| convert_constraints_bigint_to_scalar::<E, F>(c))
+        .collect();
+
+    Ok((converted_constraints, pub_io_len))
+}
+
+// Convert a set of constraints from ark-circom into R1CS format of folding_schemes
+pub fn convert_to_folding_schemes_r1cs<F>(
+    constraints: Vec<Constraints<F>>,
+    pub_io_len: usize,
+) -> R1CS<F>
+where
+    F: PrimeField,
+{
+    let mut a_matrix: Vec<Vec<(F, usize)>> = Vec::new();
+    let mut b_matrix: Vec<Vec<(F, usize)>> = Vec::new();
+    let mut c_matrix: Vec<Vec<(F, usize)>> = Vec::new();
+
+    let n_rows = constraints.len();
+
+    for (ai, bi, ci) in constraints {
+        a_matrix.push(
+            ai.into_iter()
+                .map(|(index, scalar)| (scalar, index))
+                .collect(),
+        );
+        b_matrix.push(
+            bi.into_iter()
+                .map(|(index, scalar)| (scalar, index))
+                .collect(),
+        );
+        c_matrix.push(
+            ci.into_iter()
+                .map(|(index, scalar)| (scalar, index))
+                .collect(),
+        );
+    }
+
+    let l = pub_io_len;
+    let n_cols = a_matrix.first().map(|vec| vec.len()).unwrap_or(0);
```

Comment:
This seems incorrect to me, it should be `r1cs.num_variables` iiuc.

Reply:
Thank you for your feedback. I flexed this problem.